storage/pruning/pruningStorer.go

```diff
@@ -0,0 +1,444 @@
+}
+
+// ClearCache cleans up the entire cache
+func (ps *PruningStorer) ClearCache() {
+	ps.cacher.Clear()
+}
+
+// DestroyUnit cleans up the bloom filter, the cache, and the dbs
+func (ps *PruningStorer) DestroyUnit() error {
+	ps.lock.Lock()
+	defer ps.lock.Unlock()
+
+	if ps.bloomFilter != nil {
+		ps.bloomFilter.Clear()
+	}
+
+	ps.cacher.Clear()
+
+	var err error
+	numOfPersistersRemoved := 0
+	totalNumOfPersisters := len(ps.activePersisters)
+	for _, persister := range ps.activePersisters {
+		err = persister.persister.Destroy()
+		if err != nil {
+			log.Debug("pruning db: destroy",
+				"error", err.Error())
+			continue
+		}
+		numOfPersistersRemoved++
+	}
+
+	if numOfPersistersRemoved != totalNumOfPersisters {
+		log.Debug("error destroying pruning db",
+			"identifier", ps.identifier,
+			"destroyed", numOfPersistersRemoved,
+			"total", totalNumOfPersisters)
+		return storage.ErrDestroyingUnit
+	}
+	return ps.activePersisters[0].persister.Destroy()
+}
+
+// registerHandler will register a new function to the epoch start notifier
+func (ps *PruningStorer) registerHandler(handler EpochStartNotifier) {
+	subscribeHandler := notifier.MakeHandlerForEpochStart(func(hdr data.HeaderHandler) {
+		err := ps.changeEpoch(hdr.GetEpoch())
+		if err != nil {
+			log.Warn("change epoch in storer", "error", err.Error())
+		}
+	})
+
+	handler.RegisterHandler(subscribeHandler)
+}
+
+// changeEpoch will handle creating a new persister and removing of the older ones
+func (ps *PruningStorer) changeEpoch(epoch uint32) error {
+	ps.lock.Lock()
+	defer ps.lock.Unlock()
+
+	filePath := ps.getNewFilePath(epoch)
+	db, err := ps.persisterFactory.Create(filePath)
+	if err != nil {
+		log.Warn("change epoch error", "error - "+ps.identifier, err.Error())
+		return err
+	}
+
+	newPersister := &persisterData{
+		persister: db,
+		path:      filePath,
+		isClosed:  false,
+	}
+
+	singleItemPersisters := []*persisterData{newPersister}
+	ps.activePersisters = append(singleItemPersisters, ps.activePersisters...)
+	ps.persistersMapByEpoch[epoch] = newPersister
+
+	err = ps.activePersisters[0].persister.Init()
+	if err != nil {
+		return err
+	}
+
+	err = ps.closeAndDestroyPersisters(epoch)
+	if err != nil {
+		log.Debug("closing and destroying old persister", "error", err.Error())
+		return err
+	}
+
+	return nil
+}
+
+func (ps *PruningStorer) closeAndDestroyPersisters(epoch uint32) error {
+	// recent activePersisters have to he closed for both scenarios: full archive or not
+	if ps.numOfActivePersisters < uint32(len(ps.activePersisters)) {
+		persisterToClose := ps.activePersisters[ps.numOfActivePersisters]
+		err := persisterToClose.persister.Close()
+		if err != nil {
+			log.Error("error closing persister", "error", err.Error(), "id", ps.identifier)
+			return err
+		}
+		// remove it from the active persisters slice
+		ps.activePersisters = append(ps.activePersisters[:ps.numOfActivePersisters], ps.activePersisters[ps.numOfActivePersisters+1:]...)
```

Comment:
why do you need to append older persisters? [numOfActivePersisters+1:]...

Reply:
I don't understand what you mean. This is a way to remove an element at a given position from a slice (by putting together 2 slices: one with elements before the position and one with elements after the position)

Reply:
what I meant was: aren't you supposed to always remove last? then why do you need to append anything after it?

Reply:
I understood now. fixed